tt_eager/tt_dnn/op_library/eltwise_binary/multi_core/eltwise_binary_op_multi_core.cpp

```diff
@@ -17,6 +17,221 @@ using namespace tt::constants;
+            block_height = shard_spec.value().shape[0] / TILE_HEIGHT;
+            block_width = shard_spec.value().shape[1] / TILE_WIDTH;
+            block_size = block_width * block_height;
+            end_core = (*shard_spec.value().grid.ranges().begin()).end;
+            output_width = output.get_legacy_shape()[-1] / TILE_WIDTH;
+            uint32_t output_height = output.volume() / output.get_legacy_shape()[-1] / TILE_HEIGHT;
+            last_unpadded_block_height = block_height -  (round_up(output_height, block_height) - output_height);
+            last_unpadded_block_width = block_width -  (round_up(output_width, block_width) - output_width);
+        }
+        auto bbox = core_group_1.bounding_box();
+        cores = grid_to_cores_with_noop(bbox.end.x, bbox.end.y, num_cores_x, num_cores_y, row_major);
+    } else {
+        row_major = true;
+        std::tie(num_cores, all_cores, core_group_1, core_group_2, num_tiles_per_core_group_1, num_tiles_per_core_group_2) = split_work_to_cores(compute_with_storage_grid_size, num_tiles, row_major);
+        block_cnt_per_core_group_1 = num_tiles_per_core_group_1;
+        block_cnt_per_core_group_2 = num_tiles_per_core_group_2;
+        cores = grid_to_cores(num_cores_x * num_cores_y, num_cores_x, num_cores_y, row_major);
+    }
+
+    uint32_t g1_numcores = core_group_1.num_cores();
+    uint32_t g2_numcores = core_group_2.num_cores();
+
+
+    std::vector< std::vector<uint32_t> > binary_reader_args;
+    std::vector< std::vector<uint32_t> > eltwise_binary_args;
+    std::vector< std::vector<uint32_t> > unary_writer_args;
+    if constexpr(initialize_args) {
+        binary_reader_args = { cores.size(), std::vector<uint32_t>(4)  };
+        eltwise_binary_args = { cores.size(), std::vector<uint32_t>(2)  };
+        if (block_sharded and not out_sharded)
+            unary_writer_args = { cores.size(), std::vector<uint32_t>(7) };
+        else
+            unary_writer_args = { cores.size(), std::vector<uint32_t>(3) };
+    }
+
+    auto& cached_reader_args = GetRuntimeArgs(program, binary_reader_kernel_id);
+    auto& cached_eltwise_args = GetRuntimeArgs(program, eltwise_binary_kernel_id);
+    auto& cached_writer_args = GetRuntimeArgs(program, unary_writer_kernel_id);
+    // log_info("Got cached args");
+
+    for (uint32_t i = 0, num_tiles_read = 0; i < num_cores; ++i){
+        const CoreCoord &core = cores.at(i);
+        uint32_t num_tiles_per_core = 0;
+        uint32_t block_cnt_per_core = 0;
+        uint32_t block_size_per_core = 0;
+        if (i < g1_numcores) {
+            num_tiles_per_core = num_tiles_per_core_group_1;
+            block_cnt_per_core = block_cnt_per_core_group_1;
+            block_size_per_core = block_size_per_core_group_1;
+        } else {
+            num_tiles_per_core = num_tiles_per_core_group_2;
+            block_cnt_per_core = block_cnt_per_core_group_2;
+            block_size_per_core = block_size_per_core_group_2;
+        }
```

Comment:
Wouldn't this be incorrect if `num_cores` becomes less than before when hitting cache?

Ex. Lets say we use 10 cores out of 108 cores, so initial run 10 cores have real rt args, rest of the cores are 0.

Now if we hit cache and only use 8 cores, we've set new rt args for the 8 cores, but the other 2 cores that originally had args aren't zeroed out right?

Reply:
I see.. should we iterate over all_cores and set rt args to 0 for i > num_cores?

Reply:
Does that mean that this bug is on main currently?

Reply:
> Does that mean that this bug is on main currently?

Don't think this is an issue on main because we were doing SetRuntimeArgs and updating for all cores

Reply:
I'll fix this. I believe it's only a problem if all tensors are interleaved because otherwise, shard spec is captured and num cores has to be the same.